assets/templateFloxEnv/manifest.toml

```diff
@@ -5,28 +5,42 @@
 #
 
 # List packages you wish to install in your environment inside
-# the 'install' section
-
+# the `[install]` section.
 [install]
 _FLOX_INIT_INSTALL
-# Set an environment variable.
-# These variables may not reference once another
-# and are added to the environment without expansion.
 
+# Set environment variables in the `[vars]` section.  These variables may not
+# reference once another, and are added to the environment without first
+# expanding them.  THey are available for use in the `[profile]` and `[hook]`
+# scripts.
 [vars]
 # message = "Howdy"
-# pass-in = "$some-env-var"
 
-# An activation hook will be run when entering the environment.
-# You can define one in the 'hook' table inline via the 'script' field.
+# Scripts defined in the `[profile]` section are sourced by your shell
+# immediately after entering the environment and provide a way to add shell
+# aliases to the environment or dynamically compute environment variables.  The
+# `profile.common` script is sourced by all shells, so you must write it to be
```

Comment:
I think we're providing the options to allow the user to do as they see fit for their particular use case.  If they're only using `bash` and `zsh` and doing basic stuff, they can use common just fine. If they choose to later support fish, and it breaks, then in their case, it can no longer be common and they will have to split it out to the shell specific profile sections IFF they want their environment to work on all three.

We are not defining what they put there as anything in particular... we can't since we don't know what a user will do.  We are calling it `common` to convey the idea that it *will* be sourced commonly across all shells.  It's up to the user on how they wish to use it.  In other words... it'll support whatever shells the user intends it to support, no more no less.

Reply:
I agree with Yannik that the term `common` isn't quite right here, but I think I finally understand why it is useful to have a configuration block that can be run in all cases, irrespective of the choice of userShell.

The issue boils down to this: the `hook.on-activate` only runs when the environment hasn't already been activated, but it is essential that the `profile.*` sections be sourced for subshells that a) inherit the already-activate environment and b) invoke `flox activate` from a "dotfile". **So `profile.common` can provide a place for users to define commands to be invoked with _every_ new subshell, irrespective of whether the environment has already been activated or not.**

I can see how this would be useful when activating from a "dotfile", and by putting it in the `[profile]` section of the manifest it becomes substantially more straightforward to document the following rule for "in-place" activations like this:

* when starting a new subshell which calls `flox activate` from a dotfile:
    * `[hook.on-activate]` is only sourced by `<nixShell>` if the environment isn't already active
    * `profile.common` and `profile.<userShell>` are always sourced by the `<userShell>`, in that order

On that basis I see why this is in the `[profile]` section of the config (because it pertains to the userShell and not the nixShell), but I'd almost be inclined to call it `profile.on-activate` or something like that. We should also be very clear in the template that this section is for _simple command invocations only_ and direct users to put any shell-specific language in either of the (bash-syntax) `hook.on-activate` or `profile.<userShell>` sections.

Does that help?

Reply:
> With that it seems, "common" is a bit of a misnomer considering it cant be common in the first place.

Is there a word that adequately conveys that this section can only contain a series of `<cmd> <args>` invocations that could be expected to be typed at _any_ shell prompt? Perhaps "profile.commands"? I'll be honest while I see the place in the design where it fits in, I don't know what I'd use it for so I wonder if we shouldn't be waiting to see user demand for the feature before adding it in.

Reply:
I'm looking at this more with respect to the scripts added by _our_ init helpers.
We're writing _common_ profile scripts assuming that they work in both bash and zsh.
It will just be broken for all fish users once we add support for fish that _also_ includes common (which would be kinda award as there's nothing common to fish i think).
In that case it's on them to split the script into different languages explicitly (annoying) and on us to provide 3 scripts one for each _supported_ shell (of which tow are likely identical).
The latter is kinda what python venv's are doing already...


(disclosure: just found this draft from yesterday -- its not a direct response to michael's comments above)

Reply:
I'm not sure @limeytexan 's response is a suggestion or recounting the current implementation.  If it's the latter it does not match.  @ysndr has a good point about our auto-generated scripts. So:

- `[profile]` section scripts are all *sourced*.  The `profile.common` one is always sourced, while the `profile.bash` and `profile.zsh` ones are sourced for those respective shells only.  The most basic cases doing simple things can all just go into the `common` section.  More complicated things that may be shell dependent, can go into the `common` section with the user providing the detection logic, or as a convenience, they can use the `profile.bash` and `profile.zsh` sections to avoid that logic.
- `[hook]` and the `hook.on-activate` script is executed in a bash subshell after the `profile` scripts are sourced.

As far as I can tell this is what happens on *every* `flox activate` interactive and command execution.

Anything that we place in `profile.common` is our responsibility to ensure it works in all supported shells.  If not, then perhaps we should be using the `profile.bash` and `profile.zsh` sections more.  This is a maintenance concern, when we add new shell support.  Perhaps we should add a comment at some core part of the code where we define the shells supported, and include a checklist of all the places where specific shell support is identified?

That being how things are currently working, is there a need to revisit this topic?  For nomenclature or change in functionality?

Reply:
Thanks Bill for the clarifications; one obvious issue with what you describe above is that users will want to start daemonizing processes in a non-language-specific hook section, and then be able to perform language-specific configuration based on the environment variables set by those activations, so I'm sure we will find that ordering is backwards. This is in addition to the fact that executing a hook in a subshell doesn't provide the ability to inherit env variables set by the hook, so yes I expect we'll need to revisit this at some point down the line.